Collect hydro validation errors

a-zakir opened the pull request:
collect *n* hydro validation errors (10 per area ) before exiting

flomnes committed:
Small refactor on hydro final checks (#2192)

a-zakir committed:
HydroErrorsCollector

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
fix

a-zakir committed:
logic

a-zakir committed:
update

a-zakir committed:
fix

a-zakir committed:
fix test

a-zakir committed:
fix app

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
add param

a-zakir committed:
update

a-zakir committed:
update namespace

a-zakir committed:
add include

a-zakir committed:
update member var

a-zakir committed:
update namespace

a-zakir commented on src/solver/hydro/management/HydroErrorsCollector.cpp:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    area_errors_counter_[area]++;
+    errors_limit_reached_ = area_errors_counter_[area] > 10 || errors_limit_reached_;
```
#TODO 
throw exception here?

a-zakir replied:
done

flomnes commented on src/solver/hydro/include/antares/solver/hydro/management/HydroErrorsCollector.h:
```diff
@@ -0,0 +1,43 @@
+class HydroErrorsCollector
+{
+public:
+    HydroErrorsCollector() = default;
+    void IncreaseCounterForArea(const Antares::Data::Area* area);
+    bool StopExecution() const;
+    bool ErrorsLimitReached() const;
+    void FatalErrorHit();
+    // void RecordFatalErrors(const std::string& msg, uint year);
```
That a bit too complicated for me

We should just handle limits internally, the caller should see this

```cpp
collector(area, "Error1");
collector(area, "Error2");
collector.toLog(logs.error);
```

a-zakir replied:
ok nice

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
fix

a-zakir committed:
fix

a-zakir committed:
change container

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/hydro-errors-collector

flomnes committed:
Format

flomnes commented on src/solver/hydro/management/HydroErrorsCollector.cpp:
```diff
@@ -0,0 +1,36 @@
+    errors_limit_reached_ = area_errors_counter_[area_name] > 10;
+    stop_ = true;
+
+    if (errors_limit_reached_)
+    {
+        throw FatalError("Hydro validation has failed !");
+    }
```
I think it would be best to have this behavior instead
```cpp
if (!errors_limit_reached_) {
    logs.error() << "...";
}
```
There's no need to `throw`, let higher level functions handle that. Otherwise the responsability of `HydroErrorsCollector` becomes unclear, we might even have cases where 9 errors are considered valid, but 11 are considered invalid, etc.

flomnes commented on src/solver/hydro/management/HydroInputsChecker.cpp:
```diff
@@ -50,27 +46,24 @@ void HydroInputsChecker::Execute(uint year)
     {
-        throw FatalError("hydro inputs checks: invalid minimum generation");
+        std::ostringstream msg;
+        msg << "hydro inputs checks: invalid minimum generation in year " << year;
+        errorCollector_.Collect(msg.str());
```
```suggestion
        errorCollector_ << "hydro inputs checks: invalid minimum generation in year " << year;
```
Who doesn't like a bit of syntactic sugar ?

flomnes commented on src/solver/hydro/management/HydroErrorsCollector.cpp:
```diff
@@ -0,0 +1,36 @@
+
+void HydroErrorsCollector::CheckForFatalErrors() const
+{
+    if (stop_ || errors_limit_reached_)
+    {
+        throw FatalError("Hydro validation has failed !");
+    }
```
We should `throw` if any error have been encoutered, not depending on a threshold.

a-zakir replied:
this method is called after all hydro checks (case max {errors for all area} <10)

flomnes reviewed (changes requested):
- Minor suggestion for improvement using `operator<<` overloading
- Major suggestion to remove throws, use threshold only to filter errors (the initial goal is to avoid displaying too many error messages)

Almost good !

a-zakir committed:
update

a-zakir committed:
update

flomnes commented on src/solver/hydro/include/antares/solver/hydro/management/HydroErrorsCollector.h:
```diff
@@ -0,0 +1,57 @@
+        friend void operator<<(const AreaReference& ref, const std::string& msg);
+
+        ~AreaReference()
+        {
```
Remove this (only useful for debugging)

flomnes commented on src/solver/hydro/management/HydroErrorsCollector.cpp:
```diff
@@ -0,0 +1,42 @@
+
+void HydroErrorsCollector::CheckForErrors() const
+{
+    if (!aresErrorMap_.empty())
```
```suggestion
    if (!areasErrorMap_.empty())
```

flomnes commented on src/solver/hydro/management/HydroInputsChecker.cpp:
```diff
@@ -205,4 +197,43 @@ bool HydroInputsChecker::checkGenerationPowerConsistency(uint year) const
+                                                         errorCollector_);
+          if (!validator.check())
+          {
+              logs.error() << "hydro final level : infeasibility";
```
```suggestion
              errorCollector_(area.name) << "hydro final level : infeasibility";
```

a-zakir committed:
update

a-zakir committed:
fix

a-zakir committed:
fix

a-zakir committed:
update includes

a-zakir committed:
Update src/solver/hydro/management/HydroInputsChecker.cpp

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
test

a-zakir committed:
update

a-zakir committed:
filter keys

a-zakir committed:
revert

a-zakir committed:
update

a-zakir committed:
update

flomnes committed:
Complete hydro logs (#2221)

- Instead of return, use a flag to record all possible errors
- Add precision for final level reservoir, hopefully it will make this
log message more useful

Remark : We use the newly-introduced errorCollector to limit output to
10 entries per area
